### PR TITLE
Feat/v2

### PR DESCRIPTION
**Pull request by Gozala**
Same as #11 but removes all the other code and restructures repo

### CODE REVIEW

**Review by hugomrdias — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1,53 +1,109 @@
...
+`./fauna/resources/*`, under `Collection/Index/Function` directory. File should
+be named same as a `collection/index/function` and have a `.fql` extension. E.g.
+function named `boom` would go into `./fauna/resources/Function/boom.fql`.
+
+If you need to modify a function just edit the corresponding file.
+
+If need to modify an index **think again**, most likely you should create a new
+index instead.
+
+If you need to modify a collection, please do not.
```
Comment (alanshaw) [inline README.md:70]: This seems like an unrealistic request...we'll need to change the schema at some point.

File: README.md [hunk 2]
```diff
@@ -1,53 +1,109 @@
...
+> **Note:** import-db script generates a file in `./fauna/resources/` directory for each colection/index/function. By default it will not overwrite existing files, and script will exit with code `1` when conflicting files exist. This is to avoid accidentaly overwriting changes you may have made.
+>
+> In cases where you just added new types above should be just fine. In more rare cases e.g. when you change GraphQL directives this may not be enough. In such case you will need to run `yarn import-db -- --overwrite` instead, but be aware that it **WILL OVERWRITE ALL** files in `./fauna/resources/` (becuse it just pull all defs from DB where no formatting or comments are retained). In these cases you would need to cherry-pick relevant changes and commit those in git.
+
+### Settings
+
+Code in the package assumes following environment variables which you can add to `./.env` file.
+
+- `FAUNA_KEY` - Your fauna db access token.
+- `IPFS_CLUSTER_KEY` - Access token for IPFS cluster.
```
Comment (alanshaw) [inline README.md:96]:
```suggestion
- `IPFS_CLUSTER_KEY` - HTTP basic auth token for IPFS Cluster (base64 encoded).
```

**Review by alanshaw — COMMENTED** [review]
I had viewed niftysave as a system separate from nft.storage. I'm not super comfortable with other systems pinning data to the nft.storage cluster and CIDs for NFTs being collected and stored in collections/databases that nft.storage is not directly "in charge" of.

As I mentioned in the comments, we kinda do need niftysave to use nft.storage so that data makes it into deals. Conceptually I think its easier to understand if niftysave "uses" nft.storage (and vice versa) through defined APIs. We're not super comfortable with deals data being written directly to the KV and the same feeling exists (for me at least) for niftysave writing to nft.storage's database (which I view as encompassing IPFS Cluster).

As I understand it, we're proposing storing all data for nft.storage and niftysave in the same database? What is the reason here please? I was hoping for the projects to be separate (although if there's fauna functions etc. we should code reuse obvs) for a few reasons:

1. I don't want to encourage writes to collections "owned" by other systems. This just gets messy if updates are not performed via a defined API and having all the collections in the same DB makes it easy.
2. DB maintenance to 1 system might effect the other. We can't necessarily perform maintenance on niftysave without also affecting nft.storage.
3. I don't think niftysave data written for other chains should live in the same production database as nft.storage. Every new chain will add additional load on the DB and I don't want nft.storage to be affected by it. There's also a risk of a broken github action cron accidentially hammering the DB and affecting nft.storage and lastly, we'll potentially have agencies working on other chains and I can't vouch for their ability and don't want to give more people access to the production DB than necessary.

**Comment by Gozala** [comment]
> I had viewed niftysave as a system separate from nft.storage. I'm not super comfortable with other systems pinning data to the nft.storage cluster and CIDs for NFTs being collected and stored in collections/databases that nft.storage is not directly "in charge" of.
>
>

There are few things to unpack here.

1. Database and schemas used were designed around niftysave not other way round. That said I think there is a good reasons for two systems to share same database but different APIs for working with it, I'll ellaborate on those later.
2. niftysave needed a place to pin the content and from our prior discussion my read with that we agreed to use cluster for that because nft.storage did not had an API to pin by CID. Long term I imagined we would proxy pin requests from nft.storage http endpoint so that data would end up in both cluster and pinata.

> 
> As I mentioned in the comments, we kinda do need niftysave to use nft.storage so that data makes it into deals. Conceptually I think its easier to understand if niftysave "uses" nft.storage (and vice versa) through defined APIs. We're not super comfortable with deals data being written directly to the KV and the same feeling exists (for me at least) for niftysave writing to nft.storage's database (which I view as encompassing IPFS Cluster).
> 

The impression I probably from conversations with @mikeal is that niftysave will in fact be a primary driver as in if someone pinned something on nft.storage but it never made it's way to blockchain it is a violation of terms of usage (It could be that I have misinterpreted something)

If you view system from that perspective nft.storage is more or less equivalent of preload service in IPFS, it just preloads data ahead of time.

> As I understand it, we're proposing storing all data for nft.storage and niftysave in the same database? What is the reason here please? I was hoping for the projects to be separate (although if there's fauna functions etc. we should code reuse obvs) for a few reasons:
> 
>     1. I don't want to encourage writes to collections "owned" by other systems. This just gets messy if updates are not performed via a defined API and having all the collections in the same DB makes it easy.

No writes are happening into DB directly, niftysave uses specially designed graphql endpoint which basically hands set of found tokens to the db and db does the all the processing, that includes identifying tokenURIs that aren't already in the DB (it may already be there ready if e.g. other token has same URL or e.g. if client.store API was used in nft.storage).

And then there is second CRON job that goes over all the scheduled tokenURIs and provides metadata for it, which again uses specifically designed graphql endpoint, which runs in DB and ensures both data integrity and transactional semantics. That DB endpoint will also identify all the new resources and pushes them into queue for the next job to process. Here again if resource was uploaded using nft.storage db endpoint will know to skip those.

And finally third CRON job goes over scheduler resources and either pins or download+uploads them and reports CIDs to the DB through designated GraphQL interface, which ensures integrity and constitutionality. 

I do think it makes a lot of sense to expose GraphQL endpoints to both niftysave and nft.storage positioning either of them as an owner. Transactional nature of the DB here also provides some nice benefits.

> 
>     2. DB maintenance to 1 system might effect the other. We can't necessarily perform maintenance on niftysave without also affecting nft.storage.
> 

I think it depends on kind of maintenance we are talking about here. I think a lot can be done without affecting the other, but in principal yes there will be more coordination required between multiple components if they will be sharing data base.

However benefit of this tradeoff is that:

1. It is whole a lot easier to have general consistent view of the pins etc...
2. We can directly link tokens to uploads on nft.storage and identify uploads that were not discovered on chain.
3. Avoid duplication due to inconsistencies.

It could be the tradeoffs outweigh the benefits here, but I felt it was the opposite.

>     3. I don't think niftysave data written for other chains should live in the same production database as nft.storage. Every new chain will add additional load on the DB and I don't want nft.storage to be affected by it. There's also a risk of a broken github action cron accidentially hammering the DB and affecting nft.storage and lastly, we'll potentially have agencies working on other chains and I can't vouch for their ability and don't want to give more people access to the production DB than necessary.

Current design assumed that new chains will have separate collection for tokens, but Metadata and Resource collections would be shared (think of them as DB tables). I am honestly not sure how well fauna can load-balance if writes into one collection gets a lot heavier then the other, so that indeed might be a problem.

We could allocate separate collections for each chain / service and then query and aggregate over unions of them but that would complicate things quite a bit and I am not sure it would address load-balancing concerns either.

Putting things into separate databases would make certain queries quite a bit more complicated and we would loose constitutionality but that would address load balancing concern. I am not sure however that it would necessarily address coordination issues it would just provide a bit more flex.

Overall I do not have strong opinions and if you fell like having separate DB is the way to go be it. I do not think anything needs to change for niftysave (it's just nft.storage will setup it's own separate DB) and we probably would want to expose proxy cluster API endpoints so nftifysave could pin things without having to download them.

File: README.md [hunk 3]
```diff
@@ -1,53 +1,109 @@
...
+which you can do by running `fauna-schema-migrate generate`. Unless there is a
+good reason (if so please call it out in the PR summary) generate a single
+migration (if you have generated several during development you can just delete
+them and generate new one).
+
+### GraphQL
+
+GraphQL schema is defined at `./fauna/schema.graphql`. It is used to generate
+most collections and indexes in the DB. It necessary to enable GraphQL endpoint.
+Unfortunately [fauna-schema-migrate][] currently does have GraphQL schema support.
```
Comment (alanshaw) [inline README.md:82]:
```suggestion
Unfortunately [fauna-schema-migrate][] currently does not have GraphQL schema support.
```

**Review by alanshaw — APPROVED** [review]
Lets get this merged and iterate!

File: README.md (hunk 1 above)
Comment (Gozala) [reply README.md:70]: collection just has a name really, so there's not much to modify. There are few other things that is possible to tweak e.g. like duration of history to be preserved for time based queries.

That said I agree this is unfortunate, and we could try and improve tooling to handle things better. In the meantime I think it's best to say do not change it & once we find a need to do so we'll have to improve the tooling to allow us.

File: Readme.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
+# Niftysave
```
[the remark this answers is not in the document]
Comment (Gozala) [reply Readme.md:1]: Oh that must be artifact of case sensitivity in git and host platform 😖 I just have one file on my device.

I'll try to figure it out as a followup.

**Comment by Gozala** [comment]
I was going to land it but there is one major flaw in the implementation that was identified which I would really like to be addressed before we start populating a db. When writing this I have incorrectly assumed that calling a pin on cluster would block and either respond with success when done or failure if was unable to.

@hugomrdias did point out in one of the calls that this is not the case, cluster will respond to HTTP request but perform actual pining later and update pin status accordingly. Which is why following code has a major flaw:

https://github.com/nftstorage/niftysave/blob/e2f31dbdbe01054c150dd5da00590febb8c65034/src/import-asset.js#L94-L112

Especially marking resource as pinned is problematic, because there would be a no good way to figure out which pins need rechecking and which do not.

**Comment by Gozala** [comment]
For now I'll just disable CRON jobs and land this and do a followup to address above issue and enable jobs.

**Pull request merged**